codepropertygraph/src/main/scala/io/shiftleft/utils/TimeMetric.scala

```diff
@@ -0,0 +1,349 @@
+package io.shiftleft.utils
+
+import com.sun.management.OperatingSystemMXBean
+import org.slf4j.{Logger, LoggerFactory}
+
+import java.io.PrintWriter
+import java.lang.management.ManagementFactory
+import java.util
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.{Calendar, Date}
+import scala.collection.mutable
+case class TimeMetricRecordConfig(resultFile: String = "cpuandmemoryusage.csv", recordFreq: Int = 60000)
+case class MaxPrintBuffers(MAX_FULLNAME_BUFFER: Int = 90, MAX_NAME_BUFFER: Int = 55, MAX_DONE_TIME_BUFFER: Int = 40)
+object TimeMetric {
+  val cal = Calendar.getInstance()
+  // Stage wise performance will be only recorded if parallel recording is started through initialize()
+  val stagePerformance: mutable.Map[String, NumberProcessor] = mutable.Map()
+  private val start                                          = cal.getTime
+  private var subStageStartTime                              = start
+  private var newTime                                        = start
+  private var stageStartTime                                 = start
+  private val ACROSS_PROCESS_STAGE                           = "<ACROSS_ALL>"
+  private val STAGE_NOT_SET                                  = "<not set>"
+  private val STARTED                                        = "Started"
+  private val DONE                                           = "Done"
+  private var lastStageName                                  = STAGE_NOT_SET
```

Comment:
Could these not be a stack instead? Then one could instead check if the stack is not empty and have fewer mutable variables. Also, I feel these could be combined into something like
```scala
final case class Stage(name: String, substage: mutable.Stack[String])

class TimeMetric {
// ...
  private val stages = mutable.Stack.empty[Stage]
// ...
}
```

Reply:
May also help enforce provenance if we need

Reply:
I guess a stack can be created. However, the use case that it caters to will be limited to one parent stage (group of passes), and individual passes. Introducing a stack to make it dynamic will make some logic more complicated. Let me give it a try and let you know if I face any challenges.

Reply:
I need to change the entire logic to use the stack. It will take more time to test the flow while making the changes to ensure things are working. In that case, I would probably not use the stack, as the changes are well-tested with all the combinations.